### PR TITLE
give dependent sampler access to constants

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
addresses #197

### CODE REVIEW

File: mitxgraders/sampling.py [hunk 1]
```diff
@@ -639,7 +639,7 @@ def gen_symbols_samples(symbols, samples, sample_from, functions, suffixes):
     for _ in range(samples):
         # Generate independent samples
         sample_dict = {symbol: sample_from[symbol].gen_sample() for symbol in independent}
-
+        sample_dict.update(constants)
```
Comment (jolyonb) [inline mitxgraders/sampling.py]: I think this needs to come before the previous line (well, effectively). Otherwise, you can never override a constant.

**Comment by jolyonb** [comment]
Thanks for doing this. Just the one comment.

File: mitxgraders/sampling.py (hunk 1 above)
Comment (ChristopherChudzicki) [reply mitxgraders/sampling.py]: good catch!

**Comment by jolyonb** [comment]
Hmm. Sorry, I caught one more thing. `gen_symbols_samples` now returns dictionaries containing all the variables AND constants, right? So in `FormulaGrader`'s `raw_check` function, we don't need to initialize the variables with the constants any more; we can just set the variable list to the constants.

Also, there's a cute bug here that we need to fix: a dependent variable that has the same name as a constant can lead to problems. We need to scrub the list of constants of any variable names before we do the sampling.

File: mitxgraders/sampling.py (hunk 1 above)
Comment (jolyonb) [reply mitxgraders/sampling.py]: That's why we work in pairs!

**Comment by jolyonb** [comment]
:+1:

**Comment by ChristopherChudzicki** [comment]
I think this is all set now! Good catches...

> Also, there's a cute bug here that we need to fix: a dependent variable that has the same name as a constant can lead to problems. We need to scrub the list of constants of any variable names before we do the sampling.

Took me a little while to write an example.

**Comment by jolyonb** [comment]
It's a cute little gotcha, isn't it? You'd have to be doing something fairly contrived to trigger it, but still!

**Pull request merged**